refactor(devtools): adding route details panel for selected route

sumitarora opened the pull request:
On clicking a route in the router tree, the route details panel is displayed.
- The panel shows the details of the selected route, including the path, component, providers, guards etc.
- Clicking on a chip will navigate to the corresponding source file in the editor.
- Clicking on the path will navigate to the corresponding route in the router tree.

Demo Application Source: https://github.com/sumitarora/angular-routes-demo
Demo Application Deployed: https://sumitarora.github.io/angular-routes-demo/#/


![image](https://github.com/user-attachments/assets/e9d5f4da-094d-405d-9dc2-655976c26f9a)

![image](https://github.com/user-attachments/assets/2399dfe0-3494-464c-9532-62eb880a89be)

sumitarora committed:
refactor(devtools): adding route details panel for selected route

On clicking a route in the router tree, the route details panel is displayed.
- The panel shows the details of the selected route, including the path, component, providers, guards etc.
- Clicking on a chip will navigate to the corresponding source file in the editor.
- Clicking on the path will navigate to the corresponding route in the router tree.

Demo Application Source: https://github.com/sumitarora/angular-routes-demo
Demo Application Deployed: https://sumitarora.github.io/angular-routes-demo/#/

msmallest commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.html:
```diff
@@ -1,16 +1,143 @@
+                @if (route.data.pathMatch) {
+                  <tr ng-route-details-row title="Path Match" [data]="route.data.pathMatch"></tr>
+                }
+                @if (route.data.data && route.data.data.length > 0) {
```
Would it be safe to shorten checks like these with optional chaining (`?.`)?

dgp1130 replied:
Nit: Is there any way we can rename one of these `data` properties to be a bit more clear about the distinction between them?

I know `@angular/router` already has a `data` property, and we might be stuck with that to a certain extent, but `.data.data` is hard to follow, especially when `data` is already a less-meaningful-than-most name.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+          </mat-chip>
+      } @else if (rowType() === 'flag') {
+        <span  class="{{ data() ? 'tag-active' : 'tag-incactive' }}">
+          {{ data()  ? 'Active' : 'Inactive' }}
```
Nit: Two spaces before `?`.

dgp1130 replied:
Nit: "Active" / "Inactive" I'm assuming apply to guards, but this is `rowType() === 'flag'`, and I wouldn't expect that terminology for a basic boolean. Maybe this should be `rowType() === 'guard'`?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+        color: #fff;
+      }
+    `,
+  ],
```
Style: Prefer `styles: '...'` over an array when there's only one style sheet.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+})
+export class RouteDetailsRowComponent {
+  readonly title = input<string>();
+  readonly data = input<string | string[]>();
```
Consider: Should these be `input.required`?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -216,8 +235,16 @@ export class RouterTreeVisualizer {
         return `translate(${node.x},${node.y})`;
+      })
+      .on('click', (pointerEvent: PointerEvent, node: RouterTreeD3Node) => {
+        this.nodeClickListeners.forEach((listener) => listener(pointerEvent, node));
```
Style: Prefer a native `for-of` loop over `forEach`.

https://techhub.social/@develwithoutacause/111819784287279149

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -77,5 +105,20 @@ export class RouterTreeComponent {
+    });
+  }
+
+  viewSourceFromRouter(g: string, type: string): void {
```
Consider: Is there a better name we can use than `g`? I'm assuming `c` below is supposed to be short for `component`, but I have no idea what `g` could be short for here.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -28,6 +51,11 @@ export class RouterTreeComponent {
+  private readonly _messageBus = inject(MessageBus) as MessageBus<Events>;
+  private readonly _appOperations = inject(ApplicationOperations);
+
+  selectedRoute = signal<Route | null>(null);
```
Nit: Could we `protected readonly` to constrain this a bit?

dgp1130 replied:
UX: Doesn't look like this is ever unset, would it be worth including an X in the top-right of the details panel so users can dismiss it and focus back on the graph view?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.html:
```diff
@@ -1,16 +1,143 @@
+                  <tr ng-route-details-row title="Path Match" [data]="route.data.pathMatch"></tr>
+                }
+                @if (route.data.data && route.data.data.length > 0) {
+                  <tr ng-route-details-row title="Data" [data]="route.data.data | json"></tr>
```
UX: Would it be possible to reuse the property inspector component we use on the components panel (maybe locked into read-only mode). That would potentially present much more readably.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+  ],
+})
+export class RouteDetailsRowComponent {
+  readonly title = input<string>();
```
Style: I recommend avoiding built-in attributes in directives. `<tr ng-route-details-row title="Foo">` looks like it's setting the `title` attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/title

Maybe `label` or `header` would align with the intent and avoid this correlation?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+  readonly data = input<string | string[]>();
+  readonly rowType = input<RowType>('text');
+
+  readonly chipClick = output<void | string>();
```
Nit: The name `chipClick` leaks the implementation detail that this is rendering to a `MatChip`. Maybe just `click` or `select` would be fine?

hawkgs replied:
`propertyClick`? `click` & `select` collide with the native/built-in events.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+    <th>{{title()}} :</th>
+    <td>
+      @if (rowType() === 'chip') {
+        <mat-chip (click)="chipClick.emit()">
```
Consider: Should we pass `data` through the event handler for consistency with the list case?

```html
<mat-chip (click)="chipClick.emit(data())">
  {{ data() }}
</mat-chip>
```

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+export type RowType = 'text' | 'chip' | 'flag' | 'list';
+
+@Component({
+  selector: '[ng-route-details-row]',
```
Question: Do we normally prefix attribute directives with `ng-`? Since we need to add it to `imports` anyways, I'm not sure the prefix provides much value.

sumitarora replied:
we have been using it in devtools so followed the pattern.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+  template: `
+    <th>{{title()}} :</th>
+    <td>
+      @if (rowType() === 'chip') {
```
Nit: `@switch` might be more appropriate here?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+
+export type RowType = 'text' | 'chip' | 'flag' | 'list';
+
+@Component({
```
Consider: I'm not sure this component is providing a ton of value given that it's main purpose is to switch on `rowType` and render completely different content, but then each usage of it hard-codes a specific type. It's generic, but not really used in a generic manner.

I wonder if this could be inlined in `router-tree.component.html` with each of these cases as an [`<ng-template>`](https://angular.dev/guide/templates/ng-template), then each row just picks the correct `<ng-template>` to render. This would remove an extra layer of indirection between each row and how its rendered.

It might make `router-tree.component.html` a bit large though if we could include multiple templates and the relevant styling, so up to you. Just one suggestion.

sumitarora replied:
Just wanted to have it separate if we extend it further.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+      .tag-incactive {
+        display: inline-block;
+        border-radius: 5px;
+        padding: 2px 8px;
```
A11y: Should `padding` and/or `border-radius` be `em` or `rem` units? How does this look when the user zooms in with Ctrl +?

dgp1130 replied:
I tried zooming and actually I think the padding looks fine all around.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.scss:
```diff
@@ -26,6 +26,38 @@
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  padding: 10px;
```
A11y: Similar question here, should these `padding` properties use `em` or `rem`? In my experience that usually scales better.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
Blocking: Can we include tests for at least the new code as much as we reasonably can?

I get this is modifying existing files which likely don't have much existing test coverage, but I want to make sure we make at least incremental progress on improving that.

dgp1130 reviewed (commented):
Thanks for putting this together! I know I left a lot of comments, but they're mostly pretty minor refactoring suggestions.

/cc @hawkgs in case he has any thoughts on potential UX improvements.

hawkgs commented:
LGTM for UX :+1: 

I think there are certain areas that can be improved but this can happen as part of the general effort for introducing consistency across the DevTools.

hawkgs commented on devtools/projects/ng-devtools-backend/src/lib/router-tree.ts:
```diff
@@ -102,3 +109,52 @@ function childRouteName(child: AngularRoute): string {
   }
 }
+
+export function getElementRefByName(ref: string, routes: any[], name: string): any | null {
```
`routes: any[]` => `routes: Routes` ?

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+    <th>{{title()}} :</th>
+    <td>
+      @if (rowType() === 'chip') {
+        <mat-chip (click)="chipClick.emit()">
```
+1. I think that we can just use a monospaced font instead.

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,71 @@
+
+export type RowType = 'text' | 'chip' | 'flag' | 'list';
+
+@Component({
```
_Don't want to steal Doug's conversation about whether this should be a component or a simple template hence the new comment._

Can we create separate files for the template and the styles, in case this remains as a component? I get that this comes up with some additional Bazel build configuration but it should promote better design. Right now, we won't be able to import SCSS files, for instance. This is probably rather a bigger discussion, but in my opinion, we should try to limit this pattern unless the component has a one-liner template with a single to no styles.

cc @dgp1130 @AleksanderBodurri

sumitarora replied:
done

sumitarora committed:
refactor(devtools): fixing PR comments and adding unit tests

- Update code to fix PR comments and cleanup code
- Add unit tests for the new code

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -295,10 +324,10 @@ const getProviderValue = (
     (provider) => provider.token === 'Router', // get the instance of router using token
   );
-  const routerProvider = getProviderValue(injector, routerInstance[0]);
+  const routerProvider = getProviderValue(injector, routerInstance!);
```
Bug: I was using DevTools today (actually unrelated to this PR) and noticed errors in the console because router was not on the page and `getProviderValue` was being called without a router instance.

I feel like the `!` assertion is wrong and we should return `undefined` or otherwise abort if no router could be found?

dgp1130 reviewed (commented):
Some UI feedback:

![image](https://github.com/user-attachments/assets/eac27465-db90-48ce-9f9c-bc4c837a8dd6)

* When there are too many providers it forces horizontal scrolling of the entire panel. I think we can just have those chips line-wrap and we should be good.
* `Component :` text gets a newline as the column attempts to minimize its width. Typically colons don't have spaces in front of them, so `Component:` would be more appropriate and avoid the line-wrapping issue.
* Blue boxes don't highlight when moused over.
* If a single chip is long enough to overflow, the horizontal scrollbar should probably go on the details panel rather than the entire router tree panel.
* I think I mentioned this when looking at the code, but "Active: Active" and "Auxiliary: Active" read kind of awkwardly. Should these be `true` / `false` instead (or a checkbox kind of UI)?
* "Children: 1" doesn't seem super helpful to me given that the graph is able to more visually show that. I wonder if this just more noise than it's worth in the table view?
* `feature/team` lists its `Component` as `feature/team [Lazy]` but that route doesn't actually have a component listed. I would expect this _not_ to have a component at all since it's just there to load children. Similar issue with `Component: no-name` on the root node.
* Mouseover of the details table highlights using the same color every other route is set to. I suggest either using a slightly different shade or maybe just removing highlighting on mouse hover given that the table only has two columns and I'm not sure highlighting is that useful for building a visual association of items in a row.
* At one point I managed to break mouseover highlighting of the graph nodes and it would react exactly backwards. Mousing _off_ a node would highlight it and mousing _over_ a node would remove the highlighting. Resetting DevTools fixed the issue and I haven't been able to reproduce since. So I suspect there's some bug here but I can't narrow it down any more than that.

Feature idea: The "Auxiliary" row could link to the Components panel for the associated `<router-outlet>` it's rendering to. It would be cool if we could show some meaningful name for the outlet rather than just "Auxiliary: Active", but I'm not sure such a name exists.

General UX question about the route graph more than the sidebar: Have we looked at the accessibility here for vision deficiencies? Chrome [has a feature](https://addyosmani.com/blog/emulate-vision-deficiencies-devtools/) to test out some common forms. My main takeaways are:
* Slightly thicker borders on the graph might help low contrast users.
* Deuteranopia (no green) users might struggle to distinguish active and eager loaded routes. I wonder if we can tweak the colors a bit to be more helpful here.
* Colorblind users can't distinguish the boxes at all.
    * I actually think this is mostly ok since the details panel explains textually whether a route is active or lazy and is a reasonable fallback for such users.
    * I think the main challenge here is *finding* the active route, which is difficult without color. I vaguely recall discussing making the line thicker along the path to the active route? That would probably help colorblind users find it more effectively.
* For everyone, I'm wondering if the highlight which applies from mousing over a route should also apply the same highlight to the legend in the corner? I wonder if the change in color might be more observable and help draw an association with the specific kind of route they're looking at.

Since this doesn't have too much to do with the route details panel, we can definitely follow up in a separate PR for this color accessibility stuff.

Bug in the demo app: If navigate to "Feature Team Admin" and then refresh, the navigation panel doesn't work and does not navigate anywhere.

hawkgs commented:
@dgp1130, I think that we can handle colors in a separate PR as you already pointed. The described issues with the tree visualization are already present in the injector tree since it uses the same styles. We might have to consider high contrast themes for people with vision deficiencies. 

Ps. I've already started working on the color extraction.

sumitarora commented:
- Updated PR comments and UI issues
- Fixed dark mode issues
- Changed chips to buttons for clickable links

<img width="1521" alt="image" src="https://github.com/user-attachments/assets/4098f977-ffbc-42b3-942c-545d8e33d157" />

<img width="2051" alt="image" src="https://github.com/user-attachments/assets/e71e9faf-ce72-4585-abf0-478b35c5bf26" />

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -26,6 +26,10 @@ export class RouterTreeVisualizer {
 
+  protected nodeClickListeners: ((pointerEvent: PointerEvent, node: any) => void)[] = [];
+  protected nodeMouseoverListeners: ((pointerEvent: PointerEvent, node: any) => void)[] = [];
+  protected nodeMouseoutListeners: ((pointerEvent: PointerEvent, node: any) => void)[] = [];
```
Consider: Can we type `node` as `RouterTreeD3Node`?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.html:
```diff
@@ -1,12 +1,133 @@
+          placeholder="Search routes"
+          class="filter-input"
+        />
+        <mat-checkbox (change)="togglePathSettings()">Show Full Path </mat-checkbox>
```
Typo: Trailing space.

```suggestion
        <mat-checkbox (change)="togglePathSettings()">Show Full Path</mat-checkbox>
```

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -7,20 +7,57 @@
+  {position: 8, name: 'Oxygen', weight: 15.9994, symbol: 'O'},
+  {position: 9, name: 'Fluorine', weight: 18.9984, symbol: 'F'},
+  {position: 10, name: 'Neon', weight: 20.1797, symbol: 'Ne'},
+];
```
Looks like some leftover debug data?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -29,9 +66,17 @@ export class RouterTreeComponent {
   private showFullPath = false;
 
+  private readonly _messageBus = inject(MessageBus) as MessageBus<Events>;
+  private readonly _appOperations = inject(ApplicationOperations);
```
Style: No need for leading underscores on private members, TS provides protection enough here.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.html:
```diff
@@ -0,0 +1,29 @@
+      <div class="chips-container">
+        @for (p of data(); track $index) {
+          <button mat-stroked-button (click)="click.emit(p)">
+            {{ p }}
```
Style: Is there a more meaningful name we can use than `p`? Maybe `item`?

hawkgs commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.scss:
```diff
@@ -0,0 +1,48 @@
+  display: inline-block;
+  border-radius: 5px;
+  padding: 2px 8px;
+  background: #a6d5a9;
+  color: #000000;
```
@sumitarora, sorry for adding more noise, but would you be able to transition all of the colors in the PR to the new color management system: https://github.com/angular/angular/pull/60374?

Basically, you can use any of the colors from [the scheme](https://github.com/angular/angular/blob/main/devtools/projects/ng-devtools/src/styles/_colors.scss#L56) that are similar/same to the colors in your PR. After that, you won't need to set dark colors in the stylesheets since the new mechanism will take care of that (color variable values are automatically switched on theme change).

Feel free to ping me, if something is not working well or as expected. Thanks!

sumitarora replied:
done

josephperrott reviewed (approved):
LGTM

Reviewed-for: dev-infra

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.spec.ts:
```diff
@@ -0,0 +1,126 @@
+    fixture.detectChanges();
+
+    const labelElements = fixture.debugElement.queryAll(By.css('.row-title'));
+    expect(labelElements.length).toEqual(1);
```
Consider: We probably don't need to `queryAll` and assert length 1 everywhere. `query` would likely be simpler (might want to use IDs rather than classes, but up to you).

sumitarora replied:
done

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,33 @@
+})
+export class RouteDetailsRowComponent {
+  readonly label = input.required<string>();
+  readonly data = input<any>();
```
Question: Is there a more-specific type we can use here?

sumitarora replied:
fixed

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.html:
```diff
@@ -1,12 +1,135 @@
+                  ></tr>
+                }
+
+                @if (route.data.providers &&route.data.providers.length > 0) {
```
Style: Missing space `&& route`.

sumitarora replied:
fixed

dgp1130 reviewed (commented):
Thanks for adding the tests! A couple minor comments, but nothing too concerning. If we can address the remaining open comments, I'm hopeful we can get this in pretty easily.

sumitarora commented:
@dgp1130 Fixed all comments and pushed everything.

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -166,6 +172,29 @@ const checkForAngularCallback = (messageBus: MessageBus<Events>) => () =>
 
+const navigateRouteCallback = (messageBus: MessageBus<Events>) => (path: string) => {
+  const router: any = getRouterInstance();
+  router.navigateByUrl(path);
```
We should check for navigateByUrl before calling it here

dgp1130 replied:
PR was updated to warn if `navigateByUrl` is not present.

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -373,6 +402,29 @@ export interface SerializableComponentTreeNode
+    initializeOrGetDirectiveForestHooks().getIndexedDirectiveForest(),
+    ngDebugDependencyInjectionApiIsSupported(),
+  );
+  const rootInjector = (forest[0].resolutionPath ?? []).find((i) => i.name === 'Root');
```
I wonder what we should do here for apps with multiple roots 🤔 is it even possible for more than one root to have a router configured given they'd have to share the url bar?

cc @dgp1130

sumitarora commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/route-details-row.component.ts:
```diff
@@ -0,0 +1,33 @@
+  // We need to disable the button click event for these routes.
+  readonly isRouteWithoutComponent = computed(
+    () =>
+      this.data()?.toString().includes('Lazy') || this.data()?.toString().includes('redirecting'),
```
We don't get the component reference in the parsed name of the component. That is why we need this heuristics.

sumitarora replied:
<img width="502" alt="image" src="https://github.com/user-attachments/assets/de4502e8-a137-419f-9db5-fd0f02f949ee" />

dgp1130 replied:
What exactly produces `[Lazy]`? Is that done by `ng.getRoutes` such that we can rely on its presence? Does it display to the user? I'm just wondering if there's another mechanism we should be using to represent the laziness of a route.

dgp1130 replied:
We'll defer this for now and come back if it's an issue.

sumitarora commented on devtools/projects/ng-devtools-backend/src/lib/router-tree.ts:
```diff
@@ -102,3 +145,69 @@ function childRouteName(child: AngularRoute): string {
+        return result;
+      }
+    }
+  }
```
There is already check when executing the element so even if it returns null should be good.

alan-agius4 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -383,6 +412,29 @@ export interface SerializableComponentTreeNode
+    initializeOrGetDirectiveForestHooks().getIndexedDirectiveForest(),
+    ngDebugDependencyInjectionApiIsSupported(),
+  );
+  const rootInjector = (forest[0].resolutionPath ?? []).find((i) => i.name === 'Root');
```
```suggestion
  const rootInjector = (forest[0].resolutionPath ?? []).find((i) => i.name === 'Root');
```
```suggestion
  const rootInjector = forest[0].resolutionPath?.find((i) => i.name === 'Root');
```

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/router-tree.ts:
```diff
@@ -102,3 +145,69 @@ function childRouteName(child: AngularRoute): string {
+ * @returns - the element reference if found, otherwise null
+ */
+export function getElementRefByName(
+  type: RoutePropertyType,
```
Decided to defer this until after the router types cleanup.

kirjs commented:
This PR was merged into the repository by commit 582fedd7263129b8dd11c80c053b28204d610c51.

The changes were merged into the following branches: main
